src/serlio/prtMaterial/MaterialNameSource.cpp

```diff
@@ -0,0 +1,35 @@
+#include "MaterialNameSource.h"
+
+#include "util/MItDependencyNodesWrapper.h"
+#include "util/MayaUtilities.h"
+
+#include "maya/MItDependencyNodes.h"
+
+MaterialNameSource::MaterialNameSource() {
+	MStatus status;
+	MItDependencyNodes itDependencyNodes(MFn::kShadingEngine, &status);
+	MCHECK(status);
+	for (const auto& dependencyNode : MItDependencyNodesWrapper(itDependencyNodes)) {
+		MFnDependencyNode fnDependencyNode(dependencyNode, &status);
+		MCHECK(status);
+		MString shaderNodeName = fnDependencyNode.name(&status);
+		MCHECK(status);
+		names.insert(shaderNodeName.asWChar());
+	}
+}
+
+std::pair<std::wstring, std::wstring> MaterialNameSource::getUniqueName(const std::wstring& baseName) {
+	std::wstring shadingEngineName = baseName + L"Sg";
+
+	int shIdx = 0;
+	while (names.find(shadingEngineName + std::to_wstring(shIdx)) != names.end()) {
+		shIdx++;
+	}
+	shadingEngineName += std::to_wstring(shIdx);
+	names.insert(shadingEngineName);
+
+	const std::wstring shaderName = baseName + L"Sh" + std::to_wstring(shIdx);
+	// TODO: we should also check that this shader does not exist
```

Comment:
I think the best way to go here, is to rely on the unique node names provided by Maya. I.e. in `synchronouslyCreateShadingEngine` we always create a shading group with the same name and use the return value of the MEL command `sets`, which is made unique by Maya.

Same for the shader node, let Maya choose the unique name.

Reply:
makes sense, i've removed the name source again and extended the mel script builder accordingly

Reply:
the node name handling is really nice now 👍